src/rdkafka_broker.c

```diff
@@ -4206,73 +4208,115 @@ static int rd_kafka_broker_thread_main (void *arg) {
-			 * Only the host lookup is blocking here. */
-			if (rd_kafka_broker_connect(rkb) == -1) {
-				/* Immediate failure, most likely host
-				 * resolving failed.
-				 * Try the next resolve result until we've
-				 * tried them all, in which case we sleep a
-				 * short while to avoid busy looping. */
-				if (!rkb->rkb_rsal ||
-                                    rkb->rkb_rsal->rsal_cnt == 0 ||
-                                    rkb->rkb_rsal->rsal_curr + 1 ==
-                                    rkb->rkb_rsal->rsal_cnt)
+                        /*
+                        * SASL/OAUTHBEARER is unable to connect unless a valid token is
+                        * available, and a valid token CANNOT be available unless/until
+                        * an initial token retrieval succeeds, so wait for this
+                        * precondition if necessary.
+                        */
+                        if (rkb->rkb_rk->rk_oauthbearer) {
+                                int has_token = 0;
+                                int est_total_wait_ms = 0;
+                                int max_wait_ms = 30000;
+                                while (!has_token && est_total_wait_ms < max_wait_ms) {
+                                        rwlock_rdlock(&rkb->rkb_rk->rk_oauthbearer->refresh_lock);
+                                        has_token = rkb->rkb_rk->rk_oauthbearer->token_value != NULL;
+                                        rwlock_rdunlock(&rkb->rkb_rk->rk_oauthbearer->refresh_lock);
+                                        if (has_token) {
+                                                rd_rkb_dbg(rkb, BROKER, "BRKMAIN",
+                                                        "Main broker thread has a token");
+                                        } else {
+                                                if (est_total_wait_ms >= max_wait_ms) {
+                                                        initial_token_unavailable = 1;
+                                                        rd_rkb_dbg(rkb, BROKER, "BRKMAIN",
+                                                                "Main broker thread will exit due to no initial token after %i ms",
+                                                                max_wait_ms);
+                                                } else {
+                                                        rd_rkb_dbg(rkb, BROKER, "BRKMAIN",
+                                                                "Main broker thread waiting up to %i ms for initial token",
+                                                                max_wait_ms);
+                                                        rd_kafka_broker_serve(rkb, rd_kafka_max_block_ms);
+                                                        est_total_wait_ms += rd_kafka_max_block_ms;
+                                                        if (est_total_wait_ms >= max_wait_ms) {
```

Comment:
Not necessary, better to `break` and let the next invocation of the fsm check it, it is imminent anyway and will also handle state changes (e.g., termination) correctly, otherwise the broker thread can get stuck in this loop.

Reply:
👍